plugins/processors/ifname/ifname.go

```diff
@@ -0,0 +1,328 @@
+
+func (d *IfName) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *IfName) Description() string {
+	return "Add a tag of the network interface name looked up over SNMP by interface number"
+}
+
+func (h *IfName) Init() error {
+	var err error
+	h.ifTable, err = makeTable("IF-MIB::ifTable")
+	if err != nil {
+		return err
+	}
+	h.ifXTable, err = makeTable("IF-MIB::ifXTable")
+	if err != nil {
+		return err
+	}
+
+	c := NewLRUCache(h.CacheSize)
+	h.cache = &c
+
+	return nil
+}
+
+func (d *IfName) addTag(metric telegraf.Metric) {
+	//this is called by the pool but locking is handled inside getMap
+	// if metric == nil {
+	// 	return
+	// }
+
+	agent, ok := metric.GetTag(d.AgentTag)
+	if !ok {
+		//agent tag missing
+		return
+	}
+
+	num_s, ok := metric.GetTag(d.SourceTag)
+	if !ok {
+		//source tag missing
+		return
+	}
+
+	num, err := strconv.ParseUint(num_s, 10, 64)
+	if err != nil {
+		//source tag not a uint
+		d.Log.Infof("couldn't parse source tag as uint")
+		return
+	}
+
+	m, err := d.getMap(agent)
+	if err != nil {
+		d.Log.Infof("couldn't retrieve the table of interface names: %w", err)
+		return
+	}
+
+	name, found := (*m)[num]
+	if !found {
+		//interface num isn't in table
+		d.Log.Infof("interface number %d isn't in the table of interface names", num)
+		return
+	}
+	metric.AddTag(d.DestTag, name)
+}
+
+func (d *IfName) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
+	for _, metric := range metrics {
+		d.addTag(metric)
+	}
+	return metrics
+}
+
+func (d *IfName) Start(acc telegraf.Accumulator) error {
+	d.acc = acc
+
+	fn := func(m telegraf.Metric) []telegraf.Metric {
+		d.addTag(m)
+		return []telegraf.Metric{m}
+	}
+
+	if d.Ordered {
+		d.parallel = parallel.NewOrdered(acc, fn, 10000, d.MaxParallelLookups)
+	} else {
+		d.parallel = parallel.NewUnordered(acc, fn, d.MaxParallelLookups)
+	}
+	return nil
+}
+
+func (d *IfName) Add(metric telegraf.Metric, acc telegraf.Accumulator) error {
+	d.parallel.Enqueue(metric)
+	return nil
+}
+
+func (d *IfName) Stop() error {
+	d.parallel.Stop()
+	return nil
+}
+
+// getMap gets the interface names map either from cache or from the SNMP
+// agent
+func (d *IfName) getMap(agent string) (*nameMap, error) {
+	// Check cache first
+	d.rwLock.RLock()
+	m, ok := d.cache.Get(agent)
+	d.rwLock.RUnlock()
+	if ok {
+		return m, nil
+	}
+
+	// The cache missed so make an SNMP request.  Write the response
+	// to cache and return it
+	d.rwLock.Lock()
+	defer d.rwLock.Unlock()
+
+	// Check cache again while holding write lock to prevent duplicate
+	// SNMP requests.
+	var err error
+	m, ok = d.cache.Get(agent)
+	if ok {
+		return m, nil
+	}
+
+	m, err = d.getMapRemote(agent)
+	if err != nil {
+		return nil, err
+	}
+
+	d.cache.Put(agent, m)
+	return m, nil
+}
+
+func (d *IfName) getMapRemote(agent string) (*nameMap, error) {
+	gs, err := snmp.NewWrapper(d.ClientConfig, agent)
+	if err != nil {
+		//can't translate toml to gosnmp.GoSNMP
+		return nil, err
+	}
```

Comment:
Could we do this in Start so that we can report the configuration error?  Seems like not knowing the agent might be a problem.

Reply:
I split the initialization so it only parses the host tag here and it parses the config in Start.  Now it will tell the user if any toml values are not the expected type or format.  Any snmp errors will still happen here.